contributors/design-proposals/storage/online-grow-volume-size.md

```diff
@@ -0,0 +1,344 @@
+
+The `volumeFSResizingAnnotation` annotation's key is `volumes.kubernetes.io/fs-resizing-{volumeName}`, and set it's value to `yes` if online file system resizing if required. We can use `volume.Spec.Name()` as the `volumeName` to guarantee that:
+
+* Annotation's key is a qualified name
+* `volumeName` is unique for a single pod
+
+### Prerequisite
+
+* `sig-api-machinery` need to allow pod's annotation update from kubelet. 
+* `sig-api-machinery` need to allow pod's get and annotation update from expand-controller.
+
+### Controller Manager Resize
+
+Expand controller is responsible to add the `volumeFSResizingAnnotation` to pods.
+
+#### Prerequisite of Controller Manager Resize
+
+* We use a three tuple `(pod,volume,node)` to represent a relationship: `pod` is using `volume`, `pod` is running on `node` and `volume` has been mounted to `node`. Add the `volumeFSResizingAnnotation` to `pod` only if this relationship was established.
+
+* For a specific `(volume,node)`, if more than one `pod` satisfies the relationship above, only add the annotation to one pod inside them.
+
+#### Fetch All Pods Using a PVC
+
+For a volume with online resizing request, we need to find which pods and nodes satisfy the relationship defines above, but `ExpandController` doesn't cache this relationship, so we need to fetch this information from `DesiredStateOfWorld` and `ActualStateOfWorld` of `AttachDetachController`. `DesiredStateOfWorld`caches the used volumes for each pods, `ActualStateOfWorld` caches the mounted volumes for each nodes.
+
+**Note**: `ActualStateOfWorld` only caches the mount relationship between volumes and nodes, not the volumes and pods, because only kubelet knows which pods actually mount which volumes.
+
+ To achieve this goal, we need to:
+
+* reflector the `pkg/controller/volume/attachdetach/cache` package to `pkg/controller/volume/cache`, and remove the parameter `volumePluginMgr` of `NewDesiredStateOfWorld` and `NewActualStateOfWorld` function, so that we can create a new `DesiredStateOfWorld` and `ActualStateOfWorld` without any dependence. 
+
+* This modification is possible because `DesiredStateOfWorld` and `ActualStateOfWorld` only use `volumePluginMgr` to fetch volume name, inside the `DesiredStateOfWorld.AddPod` and `ActualStateOfWorld.AddVolumeNode` method, we can get the volume name outside and transform to them as a method parameter.
+
+* Then we can create a common `DesiredStateOfWorld` and `ActualStateOfWorld`, implemented as fields of the `ControllerContext` object, instead of creating them inside `NewAttachDetachController`. After this, `AttachDetachController` and `ExpandController` can both access it through `ControllerContext`.
+
```

Comment:
Is there any example of this design being used previously? I am just curious.

Reply:
> Then we can create a common `DesiredStateOfWorld` and `ActualStateOfWorld`, implemented as fields of the `ControllerContext` object

no... controller context is not intended to be a shared mutable struct

Reply:
Lets list out all the ideas we discussed in Github issue, so as we know all the available options. The other ways of implementing this is:

1. We discussed volumemanager on kubelet side could watch for PVCs and perform FS resize when certain condition is met.  The downside of this approach is - in a large enough cluster (think 5000 nodes), it may not be good idea to create such a large number of watches. Having said that, volumemanager need not watch for all PVCs - it can only watch for PVCs which are in its actual_state_of_world. I am not sure, performance penalty of watching all PVCs VS watching only a small subset of PVCs.
2. Another option is, ExpandController can build its own object graph of pvc/pod/node mapping. This will enable us to determine pod/node where PVC is being used without sharing a cache with Attach/Detach controller.
3. We have kicked around idea of adding node name to the PVC for awhile now after attach is done. This will require an API change, but it will enable us to determine pvc->node relationship rather quickly.